studio/src/rte/rte-link-editor.js

```diff
@@ -0,0 +1,293 @@
+            ${this.#linkHrefField} ${this.#checkoutParametersField}
+            <sp-field-label for="linkText">Link Text</sp-field-label>
+            <sp-textfield
+                id="linkText"
+                placeholder="Display text"
+                .value=${this.text}
+                @input=${(e) => (this.text = e.target.value)}
+            ></sp-textfield>
+
+            <sp-field-label for="linkTitle">Title</sp-field-label>
+            <sp-textfield
+                id="linkTitle"
+                placeholder="link title"
+                .value=${this.title}
+                @input=${(e) => (this.title = e.target.value)}
+            ></sp-textfield>
+
+            <sp-field-label for="linkTarget">Target</sp-field-label>
+            <sp-picker
+                id="linkTarget"
+                .value=${this.target}
+                @change=${(e) => (this.target = e.target.value)}
+            >
+                <sp-menu>
+                    <sp-menu-item value="_self">Same Window</sp-menu-item>
+                    <sp-menu-item value="_blank">New Window</sp-menu-item>
+                    <sp-menu-item value="_parent">Parent Frame</sp-menu-item>
+                    <sp-menu-item value="_top">Top Frame</sp-menu-item>
+                </sp-menu>
+            </sp-picker>
+            ${this.#linkVariants}
+            <sp-button
+                id="cancelButton"
+                slot="button"
+                variant="secondary"
+                treatment="outline"
+                @click=${this.#handleClose}
+                type="button"
+                >Cancel</sp-button
+            >
+            <sp-button
+                id="saveButton"
+                slot="button"
+                variant="accent"
+                @click=${this.#handleSave}
+                >Save</sp-button
+            >
+        </sp-dialog>`;
+    }
+
+    get #asDialog() {
+        return html`
+            <sp-underlay ?open=${this.open}></sp-underlay>
+            ${this.#editor}
+        `;
+    }
+
+    render() {
+        if (this.dialog) return this.#asDialog;
+        return this.#editor;
+    }
+
+    #handleSave(e) {
+        e.preventDefault();
+        const data = {
+            href: this.href,
+            text: this.text,
+            title: this.title,
+            target: this.target,
+            variant: this.variant,
+        };
+        if (this.checkoutParameters !== undefined) {
+            delete data.href;
+            Object.assign(data, {
+                checkoutParameters: this.checkoutParameters,
+            });
+        }
+        this.dispatchEvent(
+            new CustomEvent('save', {
+                bubbles: true,
+                composed: true,
+                detail: data,
+            }),
+        );
+        this.#handleClose();
+    }
+
+    #handleClose() {
+        this.open = false;
+        this.dispatchEvent(
+            new CustomEvent('close', { bubbles: false, composed: true }),
```

Comment:
even though this event doesn't bubble up it might be 'safer' to use more concrete even names? like 'link-editor-close'? in future if there is some refactoring and you'll search for 'close' events you will not get 5 diff kinds of them

Reply:
I try to avoid creating very specific event names, as it makes harder for a listener to listen all of them.
SWC for instance always uses `close`, `change` events, similar to vanilla DOM.
Sometimes it is handy to listen for `close` or `change` event on a container.